flips/20210113-quorum-certificate.md

```diff
@@ -0,0 +1,153 @@
+where the collector node has been confirmed as a participant. The node uses this object to cast their raw vote.
+```cadence
+pub resource Voter {
+
+    pub let nodeID: String
+
+    // Submits the given vote. Can be called only once per epoch
+    pub fun vote(_ raw: String) {}
+}
+```
+
+The contract will also define an `Admin` resource that has the authority to create new `Voter`s, start the voting phase,
+and end the voting phase. Votes are only allowed to be submitted while the voting phase is in progress, 
+and voting is only allowed to be stopped once quorum has been reached for all clusters.
+```cadence
+pub resource Admin {
+    pub fun createVoter(nodeID: String): @Voter 
+    pub fun startVoting(clusters: [Cluster]) 
+    pub fun stopVoting()
+}
+```
+The `Admin` object will live in the Flow Service Account and be controlled by the central `FlowEpoch` contract,
+which will be detailed in another FLIP. The epoch contract will handle the complexity of generating clusters,
+restricting `Voter` creation, and timing the beginning and end of voting.
+
+The smart contract will also define public getter functions to query all of the public information in the contract,
+such as cluster information, voting completion status, etc.
+
+See this sequence diagram for the flow of actions in QC generation.
+
+[Diagram](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgRmxvdyBRdW9ydW0gQ2VydGlmaWNhdGUgR2VuZXJhdGlvbgoKCm5vdGUgcmlnaHQgb2YgRXBvY2ggQ29udHJhY3Q6IFRyaWdnZXIgRW5kIG9mIHN0YWtpbmcgQXVjADoFAB8eRGV0ZXJtaW5lIGZpbmFsIHByb3Bvc2VkIGlkZW50aXR5IHRhYmxlAB0pQ29sbGVjdG9yIENsdXN0ZXJzCgCBIg4tPlFDAIEyC0luaXRpYWxpemUgYwAyCWMANAcgYW5kIGFsbG8Agg4FVm90ZXMgdG8gbm9kZQA4H1N0YXJ0IHZvdGluZwCCIB9lbWl0AIJOBlNldHVwIEV2ZW50CgCBPgpOb2RlLT4AgmUQUmVxdWVzdACBEAVyIG9iamVjdACBWREANw46IFJldHVybgAtB2FuZCBzdG9yAIJMEAAlEGZvcm11bGF0ZQAsBnVibWkAgVUFZQCCOx5RdWVyeSByZXN1bHRzIG9mAIIFCACCcwsAgUQSVgCCLAUgY29tcGxldGVkAIMeHkVuZAArBwCCOilDb21taXR0ZWQgZQCCYgUAgV0eUHJlcGFyZQCDbAUgZm9yIG5leHQgZXBvY2gKCgo&s=default)
+
+### Drawbacks
+
+* Still researching alternative architectures. There could be better options.
+
+### Alternatives Considered
+
+* Considered deeper coupling with the other epoch contracts, 
+but this would make upgrading and unit testing more challenging
+
+### Performance Implications
+
+* checking `votingCompleted` has to iterate through all clusters and votes, which could take a lot of time and gas
+* checking `nodeHasVoted` has to iterate through all clusters and nodes, which could take a lot of time and gas,
+especially since it needs to be called by node accounts.
+* `startVoting` has to iterate through all clusters and nodes to initialize contract field each epoch, which could take a lot of time and gas. This one is harder to avoid though, but it is only called by the service account, so isn't as much of a problem.
+
```

Comment:
```suggestion
* `startVoting` has to iterate through all clusters and nodes to initialize contract field each epoch, which could take a lot of time and gas. This one is harder to avoid though, but it is only called by the service account, so isn't as much of a problem.
* including full votes in the `EpochCommit` service event results in significantly larger event size -- performing vote aggregation within the contract would make the data size of the resulting vote data proportional to the number of clusters rather than the number of collectors

```